src/main/scala/scorex/core/NodeViewHolder.scala

```diff
@@ -1,393 +1,306 @@
+  * of four elements to be updated atomically: history (log of persistent modifiers),
+  * state (result of log's modifiers application to pre-historical(genesis) state,
+  * user-specific information stored in vault (it could be e.g. a wallet), and a memory pool.
   */
 trait NodeViewHolder[TX <: Transaction, PMOD <: PersistentNodeViewModifier]
   extends Actor with ScorexLogging with ScorexEncoding {
 
   import NodeViewHolder.ReceivableMessages._
   import NodeViewHolder._
   import scorex.core.network.NodeViewSynchronizer.ReceivableMessages._
+  import scorex.core.transaction.MempoolOperation._
+  import scorex.core.transaction.state.StateOperation._
 
   type SI <: SyncInfo
-  type HIS <: History[PMOD, SI, HIS]
-  type MS <: MinimalState[PMOD, MS]
-  type VL <: Vault[TX, PMOD, VL]
-  type MP <: MemoryPool[TX, MP]
+  type History <: consensus.History[PMOD, SI, History]
+  type State <: MinimalState[PMOD, State]
+  type Vault <: transaction.wallet.Vault[TX, PMOD, Vault]
+  type MPool <: MempoolReader[TX]
 
-  type NodeView = (HIS, MS, VL, MP)
+  /** State actor used for asynchronous transaction validation
+    */
+  protected val stateActor: ActorRef = createStateActor()
 
-  case class UpdateInformation(history: HIS,
-                               state: MS,
-                               failedMod: Option[PMOD],
-                               alternativeProgressInfo: Option[ProgressInfo[PMOD]],
-                               suffix: IndexedSeq[PMOD])
+  /** Underlying memory pool actor
+    */
+  protected val memoryPoolActor: ActorRef = createMemoryPoolActor()
+
+  protected var vault: Vault = restoreVault().getOrElse(genesisVault())
 
   val scorexSettings: ScorexSettings
 
-  /**
-    * Cache for modifiers. If modifiers are coming out-of-order, they are to be stored in this cache.
-    */
-  protected lazy val modifiersCache: ModifiersCache[PMOD, HIS] =
-    new DefaultModifiersCache[PMOD, HIS](scorexSettings.network.maxModifiersCacheSize)
-
-  /**
-    * The main data structure a node software is taking care about, a node view consists
-    * of four elements to be updated atomically: history (log of persistent modifiers),
-    * state (result of log's modifiers application to pre-historical(genesis) state,
-    * user-specific information stored in vault (it could be e.g. a wallet), and a memory pool.
-    */
-  private var nodeView: NodeView = restoreState().getOrElse(genesisState)
+  protected implicit val defaultAskTimeout: Timeout = Timeout(10.seconds)
+  protected implicit val executionContext: ExecutionContext = context.dispatcher
 
-  /**
-    * Restore a local view during a node startup. If no any stored view found
+  /** Restore a local history during a node startup. If no any stored view found
     * (e.g. if it is a first launch of a node) None is to be returned
     */
-  def restoreState(): Option[NodeView]
+  protected def restoreHistory(): Option[History]
```

Comment:
Shouldn't this method be moved in `StateHistoryActor` now, as there are no state and history variables now in NVH?

Reply:
Now I think, you are right, and these should be moved to the correspondent actors